### ISSUE TITLE
2.x: Cairo backends cannot render images

### ISSUE THREAD

**Issue by QuLogic** [matplotlib/matplotlib#6562]
In testing #6540, I can see that neither Cairo backend is able to render images correctly.

With GTKCairo:
```pytb
Traceback (most recent call last):
  File ".../lib/matplotlib/backends/backend_gtk.py", line 438, in expose_event
    self._render_figure(self._pixmap, w, h)
  File ".../lib/matplotlib/backends/backend_gtk.py", line 426, in _render_figure
    self.figure.draw (self._renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/figure.py", line 1129, in draw
    renderer, self, dsu, self.suppressComposite)
  File ".../lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/axes/_base.py", line 2353, in draw
    mimage._draw_list_compositing_images(renderer, self, dsu)
  File ".../lib/matplotlib/image.py", line 163, in _draw_list_compositing_images
    flush_images()
  File ".../lib/matplotlib/image.py", line 147, in flush_images
    image_group[0].draw(renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/image.py", line 464, in draw
    renderer.draw_image(gc, l, b, im)
  File ".../lib/matplotlib/backends/backend_cairo.py", line 180, in draw_image
    im.shape[1]*4)
TypeError: expected a writeable buffer object
```

With GTK3Cairo:
```pytb
Traceback (most recent call last):
  File ".../lib/matplotlib/backends/backend_gtk3cairo.py", line 45, in on_draw_event
    self._render_figure(w, h)
  File ".../lib/matplotlib/backends/backend_gtk3cairo.py", line 34, in _render_figure
    self.figure.draw (self._renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/figure.py", line 1129, in draw
    renderer, self, dsu, self.suppressComposite)
  File ".../lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/axes/_base.py", line 2353, in draw
    mimage._draw_list_compositing_images(renderer, self, dsu)
  File ".../lib/matplotlib/image.py", line 163, in _draw_list_compositing_images
    flush_images()
  File ".../lib/matplotlib/image.py", line 147, in flush_images
    image_group[0].draw(renderer)
  File ".../lib/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File ".../lib/matplotlib/image.py", line 464, in draw
    renderer.draw_image(gc, l, b, im)
  File ".../lib/matplotlib/backends/backend_cairo.py", line 180, in draw_image
    im.shape[1]*4)
TypeError: expected a writeable buffer object
```

Bisect points to 659513951920d83fe6ef68ec40dd72f7bd6d6653, the backport of #5718.

**Comment by QuLogic** [comment]
Bisect on the original branch points to one of c9cae50246ac02cc992a27a604461edd65a5b8dc 3b7d1bb73f73ca667f3978e9d8201112761d6ee9 4a0530237a6df7b92ba0aba36d27e634a2eeab55 0aa222a647e9b5f2ed882afb7fe78fab3eebb28b cf11aea8172be914cc96dcf3329bf7b694ba1d27 0793aa2fe6ce79eb265d4c4a2d1d5494061e59e9; perhaps that's narrow enough to figure it out.

cc @mdboom

**Comment by jenshnielsen** [comment]
Interestingly enough this seems to be a python2 issue only. I can reproduce the error on python 2.7 but not on 3.5

The following reproduces the error without matplotlib

```python
try:
    import cairocffi as cairo
except:
    import cairo
import numpy as np
im = np.random.rand(100,100,4)
cairo.ImageSurface.create_for_data(memoryview(im.flatten()),
                                   cairo.FORMAT_ARGB32,
                                   im.shape[1],
                                   im.shape[0], im.shape[1]*4)
```

**Comment by jenshnielsen** [comment]
So a in cairocffi this is fundamentally because `ctypes.c_char.from_buffer_copy(memoryview(data))`
doesn't work. Regardless of the type of data. Numpy array, bytesarray ...

**Comment by jenshnielsen** [comment]
This stackoverflow question is an example of the same issue http://stackoverflow.com/questions/28984692/ctypes-from-buffer-with-memoryviews-in-python-2-7-and-python-3-4

**Comment by QuLogic** [comment]
Fixed by #6602.
